### ISSUE TITLE
FTP upload on save

### ISSUE THREAD

**Issue by jarodium** [c9/core#65]
Hello, 

When a ftp mount is configured, I should be able to automatically upload ( via preference setting ) the file i'm working on, like auto-sync.

On the other hand, there could be another option or method which would mark the files I just saved, and by pressing a key combination, upload the files marked.

**Comment by pegasussan** [comment]
+1
That would be VERY usefull...

**Comment by javruben** [comment]
FTP mounting, creates a file system mount where each file is immediately saved on the remote server. Have you got this to work? What is it that you are looking for exactly?

**Comment by jarodium** [comment]
I have some local files. 
Upon save, I would like to upload them to the FTP server.
From what I can see, the fs mount on the server, allows me to work on the server files directly, but I can't maintain a synced version without having to download all from the server.

**Comment by javruben** [comment]
Still not 100% sure.

FTP mounting does *not* download the files. Any file you save will be immediately saved on the remote server. If I understand correctly, this is what you are looking for...

**Comment by jarodium** [comment]
I will test it properly on the weekend. And I will let you know... 
Maybe I still have an older version of the SDK...

**Comment by pegasussan** [comment]
I believe I can clarify a little bit.
Both of you seem right from what I have tested.
If you work with remote files on FTP, they will be saved remotely. But what jarodium seems to be talking about is the ability to work locally and have cloud nine save the same file locally and remotely at the same time. Something similar to Netbeans "sync on save" ability.
That way one could work on local projects, synced with remote ones.
Am I correct jarodium??

**Comment by jarodium** [comment]
@pegasussan yes. that is correct.

**Comment by javruben** [comment]
Aha! Thanks @pegasussan.

We had a feature like that once, using lftp in the background. It didnt do automatic syncing though. You had to do something to sync.

@xoob Could your syncing solution be generalized for other sync methods, such as syncing over ftp?

**Comment by xoob** [comment]
@javruben Yes. This could definitely be turned into a generic sync solution with different backends. I could imagine FTP, S3, etc.

**Comment by jarodium** [comment]
@xoob is there any progress in the FTP syncing?

**Comment by xoob** [comment]
@jarodium No; we have it on the roadmap but there's no concrete target date yet for adding this feature.

**Comment by peterf1972** [comment]
I would love that function too. I had been using Netbeans before switching to C9, and Netbeans also has the function to upload files to an attached ftp directory. You can choose 'automatically: save on run' or save manually (right click on file, then 'upload').

**Comment by nightwing** [comment]
Unfortunately it is not clear when we will be able to implement this feature, but if someone wants to implement it as a plugin i'd be happy to help.

If you want only to upload on save, you can use a simple init script like this:

```js
var dirname = require("path").dirname
var c9 = services.c9

var user = "username"
var password = "password"
var url = "sftp://my-host:22"
var remoteRoot = "/WWW"
var localRoot = c9.workspaceDir

function lftpSave(path) {
    services.proc.execFile("lftp", {args: [
        "-c",
        `open -u ${user},${password} ${url}; mkdir -p "${dirname(remoteRoot + path)}"; put "${localRoot}${path}" -o "${remoteRoot}${path}"`
    ]}, function(err,stdout,stderr) {
        console.log(stdout, stderr)
    })
}
// if you want to upload automatically
services.save.on("afterSave", function(e) { 
    var path = e.path
    lftpSave(path)
})
```
